win32ss/user/conime/conime.c

```diff
@@ -0,0 +1,1627 @@
+                pEntry->adwData[i] = 0;
+            }
+        }
+
+        if (pEntry->pLocal2)
+        {
+            LocalFree(pEntry->pLocal2);
+            pEntry->unknown3_5[0] = 0;
+        }
+
+        if (pEntry->pKLInfo)
+            pEntry->pKLInfo = LocalFree(pEntry->pKLInfo);
+
+        LocalFree(pEntry);
+    }
+
+    g_cEntries = 0;
+    g_pUserData = LocalFree(g_pUserData);
+
+    RtlLeaveCriticalSection(&g_csLock);
+
+    if (g_dwAttachToThreadId && (uMsg != WM_ENDSESSION || !g_bIsLogOnSession))
+    {
+        /* Detach the thread input */
+        AttachThreadInput(GetCurrentThreadId(), g_dwAttachToThreadId, FALSE);
+
+        /* Unregister the Console IME */
+        UnregisterConsoleIME();
+
+        g_dwAttachToThreadId = 0;
+    }
+}
+
+/* WM_ENABLE */
+LRESULT ConIme_OnEnable(HWND hWnd, WPARAM wParam, LPARAM lParam)
+{
+    UINT i;
+    PENTRY pEntry;
+
+    if (wParam)
+    {
+        RtlEnterCriticalSection(&g_csLock);
+        for (i = 1; i < g_cEntries; ++i)
+        {
+            pEntry = g_pUserData->pEntries[i];
+            if (pEntry)
+            {
+                if (pEntry->bEnabled)
+                {
+                    if (!pEntry->unknown3_5[3] && !IsWindowEnabled(pEntry->hwndConsole))
+                    {
+                        EnableWindow(pEntry->hwndConsole, TRUE);
+                        pEntry->unknown3_5[3] = 1;
+                        if (!pEntry->bWndEnabled)
+                            SetForegroundWindow(pEntry->hwndConsole);
+                    }
+                }
+            }
+        }
+        RtlLeaveCriticalSection(&g_csLock);
+    }
+    else
+    {
+        pEntry = IntGetEnabledEntry(g_bEnabled);
+        if (pEntry)
+        {
+            if (pEntry->bEnabled)
+            {
+                pEntry->unknown3_5[3] = 0;
+                EnableWindow(pEntry->hwndConsole, 0);
+                FIXME("\n");
+            }
+        }
+    }
+
+    return DefWindowProcW(hWnd, WM_ENABLE, wParam, lParam);
+}
+
+/* @implemented : WM_KEYFIRST ... WM_KEYLAST */
+BOOL ConIme_OnKeyChar(HWND hWnd, UINT uMsg, WPARAM wParam, LPARAM lParam)
+{
+    PENTRY pEntry = IntGetEnabledEntry(g_bEnabled);
+    if (!pEntry)
+        return FALSE;
+    return PostMessageW(pEntry->hwndConsole, uMsg + 0x800, wParam, lParam); /* ==> 0x900... */
+}
+
+/* @implemented : IMN_OPENSTATUSWINDOW */
+BOOL ConIme_UpdateImeState(HWND hWnd)
+{
+    PENTRY pEntry;
+    HIMC hIMC;
+    PVOID lpData;
+    COPYDATASTRUCT CopyData;
+
+    pEntry = IntGetEnabledEntry(g_bEnabled);
+    if (!pEntry)
+        return FALSE;
+
+    hIMC = ImmGetContext(hWnd);
+    if (!hIMC)
+        return FALSE;
+
+    lpData = LocalAlloc(LPTR, 0x288);
+    if (!lpData)
+    {
+        ImmReleaseContext(hWnd, hIMC);
+        return FALSE;
+    }
+
+    ImmGetConversionStatus(hIMC, &pEntry->dwConversion, &pEntry->dwSentence);
+
+    CopyData.dwData = CONIME_SEND_CONV_MODE;
+    CopyData.cbData = 0x288;
```

Comment:
why? What is that hardcoded value equal to?

Reply:
Fixed as `sizeof(IME_STATUS)`.